src/Datadog.Trace/DuckTyping/DuckTypeExtensions.cs

```diff
@@ -41,10 +41,14 @@ public static object DuckCast(this object instance, Type targetType)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static bool TryDuckCast<T>(this object instance, out T value)
         {
-            if (DuckType.CanCreate<T>(instance))
+            if (instance is not null)
             {
-                value = DuckType.Create<T>(instance);
-                return true;
+                DuckType.CreateTypeResult proxyResult = DuckType.CreateCache<T>.GetProxy(instance.GetType());
+                if (proxyResult.Success)
+                {
+                    value = proxyResult.CreateInstance<T>(instance);
+                    return true;
+                }
```

Comment:
I understand it may be for optimisation reasons, but could we not reduce the duplication in these extensions? for example:

```csharp
public static bool TryDuckCast<T>(this object instance, out T value)
    => instance.TryDuckCast(typeof(T), value);

public static bool DuckAs<T>(this object instance)
    => instance.TryDuckCast<T>(out var value) ? value : null;

public static bool DuckAs<T>(this object instance, Type targetType)
    => instance.TryDuckCast(targetType, out var value) ? value : null;
```

Reply:
Yeah If I do that, we remove the `CreateCache<T>` optimization, so it will perform a bit worse. 

static field vs dictionary lookup.

Reply:
oh yeah, I missed that.  Ignore me 🙂